e2e/LongHaul/service/IotHub.cs

```diff
@@ -124,78 +71,37 @@ public async Task MonitorConnectedDevicesAsync(CancellationToken ct)
 
                 await foreach (ClientTwin device in allDevices)
                 {
-                    if (s_activeDeviceSet.Contains(device.DeviceId) && device.ConnectionState == ClientConnectionState.Disconnected)
+                    string deviceId = device.DeviceId;
+
+                    if (s_onlineDeviceOperations.ContainsKey(deviceId) && device.ConnectionState is ClientConnectionState.Disconnected)
                     {
-                        s_activeDeviceSet.Remove(device.DeviceId);
+                        s_onlineDeviceOperations.Remove(deviceId);
                     }
-                    else if (!s_activeDeviceSet.Contains(device.DeviceId) && device.ConnectionState == ClientConnectionState.Connected)
+                    else if (!s_onlineDeviceOperations.ContainsKey(deviceId) && device.ConnectionState is ClientConnectionState.Connected)
                     {
-                        s_activeDeviceSet.Add(device.DeviceId);
+                        s_onlineDeviceOperations.Add(
+                            deviceId,
```

Comment:
How do you cancel the running of these tasks for this device but not the rest of the app?

I'd suggest having device-specific cancellation token sources, so when a device disconnects, you can signal it and remove it from this list.